File: cpp/okf/validate.cpp
// OKF validation rules — mirrors py/okf/okf.py::validate verbatim.
// Permissive consumption: recommended-field issues are warnings; only
// unparseable frontmatter / missing type are errors.

#include <regex>

#include "okf.hpp"

namespace okf {

std::vector<Finding> validate(const Bundle& b) {
    static const std::regex re_iso(R"(^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$)");
    std::vector<Finding> out;
    auto add = [&out](const std::string& path, const char* sev, const char* rule,
                      const std::string& msg) {
        out.push_back(Finding{path, sev, rule, msg});
    };

    for (const Concept& c : b.concepts) {
        if (c.reserved) continue;
        if (c.parse_error) {
            add(c.path, "error", "frontmatter_unparseable",
                "no parseable frontmatter (" + *c.parse_error + ")");
            continue;
        }
        if (!c.type || c.type->empty()) {
            add(c.path, "error", "missing_type", "frontmatter has no non-empty type");
        }
        if (!c.title) add(c.path, "warn", "missing_title", "recommended field title absent");
        if (!c.description) {
            add(c.path, "warn", "missing_description", "recommended field description absent");
        }
        if (!c.timestamp) {
            add(c.path, "warn", "missing_timestamp", "recommended field timestamp absent");
        } else if (!std::regex_match(*c.timestamp, re_iso)) {
            add(c.path, "warn", "timestamp_not_iso8601",
                "timestamp not ISO-8601: " + *c.timestamp);
        }
    }

    std::vector<Link> lk_all = links(b);
    for (const Link& lk : lk_all) {
        if (!lk.resolved) {
            add(lk.src_path, "warn", "broken_link", "unresolved link: " + lk.dst_raw);
        }
    }

    // orphan concepts: non-reserved, parseable, no inbound link
    std::set<std::string> inbound;
    for (const Link& lk : lk_all) {
        if (lk.dst_path) inbound.insert(*lk.dst_path);
    }
    for (const Concept& c : b.concepts) {
        if (c.reserved || c.parse_error) continue;
        if (!inbound.count(c.path)) {
            add(c.path, "warn", "orphan", "no inbound links (orphan concept)");
        }
    }
    return out;
}

}  // namespace okf

File: rust/okf-ingest/src/validate.rs
//! OKF validation rules — mirrors `py/okf/okf.py::validate` verbatim.
//! Permissive consumption: recommended-field issues are warnings, only
//! unparseable frontmatter / missing type are errors.

use std::collections::BTreeSet;
use std::sync::OnceLock;

use regex::Regex;

use crate::links::links;
use crate::model::{Bundle, Finding, Severity};

fn re_iso() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$").unwrap())
}

pub fn validate(b: &Bundle) -> Vec<Finding> {
    let mut out: Vec<Finding> = Vec::new();
    let add = |out: &mut Vec<Finding>, path: &str, sev: Severity, rule: &str, msg: String| {
        out.push(Finding {
            path: path.to_string(),
            severity: sev,
            rule: rule.to_string(),
            message: msg,
        });
    };

    for c in &b.concepts {
        if c.reserved {
            continue;
        }
        if let Some(err) = &c.parse_error {
            add(
                &mut out,
                &c.path,
                Severity::Error,
                "frontmatter_unparseable",
                format!("no parseable frontmatter ({err})"),
            );
            continue;
        }
        if c.kind.as_deref().unwrap_or("").is_empty() {
            add(
                &mut out,
                &c.path,
                Severity::Error,
                "missing_type",
                "frontmatter has no non-empty type".to_string(),
            );
        }
        if c.title.is_none() {
            add(
                &mut out,
                &c.path,
                Severity::Warn,
                "missing_title",
                "recommended field title absent".to_string(),
            );
        }
        if c.description.is_none() {
            add(
                &mut out,
                &c.path,
                Severity::Warn,
                "missing_description",
                "recommended field description absent".to_string(),
            );
        }
        match &c.timestamp {
            None => add(
                &mut out,
                &c.path,
                Severity::Warn,
                "missing_timestamp",
                "recommended field timestamp absent".to_string(),
            ),
            Some(ts) if !re_iso().is_match(ts) => add(
                &mut out,
                &c.path,
                Severity::Warn,
                "timestamp_not_iso8601",
                format!("timestamp not ISO-8601: {ts}"),
            ),
            _ => {}
        }
    }

    let lk_all = links(b);
    for lk in &lk_all {
        if !lk.resolved {
            add(
                &mut out,
                &lk.src_path,
                Severity::Warn,
                "broken_link",
                format!("unresolved link: {}", lk.dst_raw),
            );
        }
    }

    // orphan concepts (Karpathy-style lint): non-reserved, parseable, no inbound link
    let inbound: BTreeSet<&str> = lk_all
        .iter()
        .filter_map(|l| l.dst_path.as_deref())
        .collect();
    for c in &b.concepts {
        if c.reserved || c.parse_error.is_some() {
            continue;
        }
        if !inbound.contains(c.path.as_str()) {
            add(
                &mut out,
                &c.path,
                Severity::Warn,
                "orphan",
                "no inbound links (orphan concept)".to_string(),
            );
        }
    }
    out
}
